source/common/router/router.cc

```diff
@@ -462,6 +462,24 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::HeaderMap& trailers) {
   return Http::FilterTrailersStatus::StopIteration;
 }
 
+Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
+  Http::MetadataMap metadata;
+  metadata.insert(metadata_map.begin(), metadata_map.end());
+  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata);
+
+  ASSERT(upstream_requests_.size() == 1);
+  if (upstream_requests_.front() == nullptr) {
+    ENVOY_STREAM_LOG(trace,
+                     "the upstream_request_ not ready. Storing metadata_map to encode later: {}",
```

Comment:
upstream_request_ -> upstream_request_ is

Reply:
Though unless metadata can race ahead of headers, I don't think we need to worry about this.  I think we want to handle "no upstream connection" in one place.  Right now when we get headers we always create the upstream request, and then eventually when we have a stream to the backend we have an encoder.  I think it'd be good to eagerly create, and then buffer in the upstream request consistently.

Reply:
I think metadata may arrive before headers when decodeHeaders() inserts new metadata. That because addDecodedMetadata() will call decodeMetadata() through all filters. At that time, headers may only half way through the filters.

Reply:
OK, in that case what if we create the upstream request when we receive the first frame (data or metadata) and always let the upstream request buffer?

Reply:
Good idea! Fixed.